excellent/types/conversions.go

```diff
@@ -1,303 +1,138 @@
-		num, err := ToDecimal(env, asString)
+	switch typed := x.(type) {
+	case XError:
+		return XDateZero, typed
+	case XDate:
+		return typed, nil
+	case XString:
+		parsed, err := utils.DateFromString(env, typed.Native())
 		if err == nil {
-			return !num.Equals(decimal.Zero), nil
-		}
-
-		noWhite := strings.Join(strings.Fields(asString), "")
-
-		// empty array?
-		if noWhite == "[]" {
-			return false, nil
+			return NewXDate(parsed), nil
 		}
-
-		// empty dict
-		if noWhite == "{}" {
-			return false, nil
-		}
-
-		// finally just string version
-		return asString != "" && strings.ToLower(asString) != "false", nil
-
-	case Atomizable:
-		return ToBool(env, test.Atomize())
 	}
 
-	return false, fmt.Errorf("unable to convert value '%+v' of type '%s' to a bool", test, reflect.TypeOf(test))
+	return XDateZero, NewXErrorf("unable to convert value '%v' of type '%s' to a date", x, reflect.TypeOf(x))
 }
 
-// Compare returns the difference between the two passed interfaces which must be of the same type
-func Compare(env utils.Environment, arg1 interface{}, arg2 interface{}) (int, error) {
-	if arg1 == nil && arg2 == nil {
-		return 0, nil
+// ToInteger tries to convert the passed in value to an integer or returns an error if that isn't possible
+func ToInteger(x XValue) (int, XError) {
+	number, err := ToXNumber(x)
+	if err != nil {
+		return 0, err
 	}
 
-	arg1, arg1Type, _ := ToXAtom(env, arg1)
-	arg2, arg2Type, _ := ToXAtom(env, arg2)
-
-	// common types, do real comparisons
-	switch {
-	case arg1Type == arg2Type && arg1Type == XTypeError:
-		return strings.Compare(arg1.(error).Error(), arg2.(error).Error()), nil
-
-	case arg1Type == arg2Type && arg1Type == XTypeDecimal:
-		return arg1.(decimal.Decimal).Cmp(arg2.(decimal.Decimal)), nil
-
-	case arg1Type == arg2Type && arg1Type == XTypeBool:
-		bool1 := arg1.(bool)
-		bool2 := arg2.(bool)
-
-		switch {
-		case !bool1 && bool2:
-			return -1, nil
-		case bool1 == bool2:
-			return 0, nil
-		case bool1 && !bool2:
-			return 1, nil
-		}
-
-	case arg1Type == arg2Type && arg1Type == XTypeTime:
-		time1 := arg1.(time.Time)
-		time2 := arg2.(time.Time)
+	intPart := number.Native().IntPart()
 
-		switch {
-		case time1.Before(time2):
-			return -1, nil
-		case time1.Equal(time2):
-			return 0, nil
-		case time1.After(time2):
-			return 1, nil
-		}
-
-	case arg1Type == arg2Type && arg1Type == XTypeString:
-		return strings.Compare(arg1.(string), arg2.(string)), nil
+	if intPart < math.MinInt32 && intPart > math.MaxInt32 {
+		return 0, NewXErrorf("number value %s is out of range for an integer", string(number.ToString()))
 	}
 
-	if arg1Type != arg2Type {
-		return 0, fmt.Errorf("Cannot compare different types of %#v and %#v", arg1, arg2)
-	}
+	return int(intPart), nil
+}
 
-	arg1Str := fmt.Sprintf("%v", arg1)
-	arg2Str := fmt.Sprintf("%v", arg2)
+// RequireMarshalToXString calls json.Marshal in the given value and panics in the case of an error
+func RequireMarshalToXString(x interface{}) XString {
```

Comment:
I like this, though I think the golang convention is `must` instead of `require`

Reply:
can change - copied Require from `decimal.RequireFromString`

Reply:
Oh, well maybe I'm wrong, maybe worth checking what is used more, no biggie either way.

Reply:
I think you're right about Must being more standard - and `MustMarshalToXString` sounds marginally better